rts/motoko-rts/src/persistence/compatibility.rs

```diff
@@ -136,37 +138,72 @@ unsafe fn create_type_check_cache<M: Memory>(
-/// This uses the existing IDL subtype test.
-pub unsafe fn memory_compatible<M: Memory>(
-    mem: &mut M,
-    old_type: &mut TypeDescriptor,
-    new_type: &mut TypeDescriptor,
-) -> bool {
-    let cache = create_type_check_cache(mem, old_type, new_type);
-
-    let old_type_table = old_type.build_type_table(mem);
-    let old_table_end = old_type.type_table_end();
-
-    let new_type_table = new_type.build_type_table(mem);
-    let new_table_end = new_type.type_table_end();
-
-    crate::idl::memory_compatible(
-        &cache,
-        TypeVariance::Covariance,
-        old_type_table,
-        new_type_table,
-        old_table_end,
-        new_table_end,
-        MAIN_ACTOR_TYPE_INDEX,
-        MAIN_ACTOR_TYPE_INDEX,
-        true,
-    )
+// Helper structure to support multiple consecutive memory tests
+// for the same type tables.
+// This contains low-level pointers and must only be used within the same message,
+// not across GC increments.
+pub struct MemoryCompatibilityTest {
+    cache: BitRel,
+    old_type_table: *mut *mut u8,
+    old_table_end: *mut u8,
+    new_type_table: *mut *mut u8,
+    new_table_end: *mut u8,
+}
+
+impl MemoryCompatibilityTest {
+    // Build the temporary type table with absolute addresses, to use
+    // it one or multiple times within the same IC message.
+    pub unsafe fn new<M: Memory>(
+        mem: &mut M,
+        old_type: &mut TypeDescriptor,
+        new_type: &mut TypeDescriptor,
+    ) -> Self {
+        let cache = create_type_check_cache(mem, old_type, new_type);
+
+        let old_type_table = old_type.build_type_table(mem);
+        let old_table_end = old_type.type_table_end();
+
+        let new_type_table = new_type.build_type_table(mem);
+        let new_table_end = new_type.type_table_end();
+
+        Self {
+            cache,
+            old_type_table,
+            old_table_end,
+            new_type_table,
+            new_table_end,
+        }
+    }
+
+    /// Test whether the new stable type complies with the existing old stable type.
+    /// This uses the existing IDL subtype test.
+    pub unsafe fn is_compatible(&self, old_type_index: i32, new_type_index: i32) -> bool {
+        debug_assert!(
+            old_type_index != MAIN_ACTOR_TYPE_INDEX || new_type_index == MAIN_ACTOR_TYPE_INDEX
+        );
+        self.cache.init(); // Needs to be-reinitialized on every check.
+        crate::idl::memory_compatible(
+            &self.cache,
+            TypeVariance::Covariance,
+            self.old_type_table,
+            self.new_type_table,
+            self.old_table_end,
+            self.new_table_end,
+            old_type_index,
+            new_type_index,
+            new_type_index == MAIN_ACTOR_TYPE_INDEX,
+        )
+    }
+
+    // Test whether the new stable type of the main actor complies with
+    // the existing old stable type of the main actor.
+    pub unsafe fn compatible_stable_actor(&self) -> bool {
```

Comment:
Is this just comparing the stable actor type to itself? Is this just a sanity check?